domain/annotation/state/state.go

```diff
@@ -0,0 +1,274 @@
+		annotations[result.Key] = result.Value
+	}
+
+	return annotations, errors.Trace(err)
+}
+
+// SetAnnotations associates key/value annotation pairs with a given ID.
+// If annotation already exists for the given ID, then it will be updated with
+// the given value. Setting a key's value to "" will remove the key from the annotations map
+// (functionally unsetting the key).
+func (st *State) SetAnnotations(ctx context.Context, id annotations.ID,
+	annotationsParam map[string]string) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// Separate the annotations that are to be setted vs removed
+	toUpsert := make(map[string]string)
+	var toRemove []string
+
+	for key, value := range annotationsParam {
+		if value == "" {
+			toRemove = append(toRemove, fmt.Sprintf("'%s'", key))
+		} else {
+			toUpsert[key] = value
+		}
+	}
+
+	if id.Kind == annotations.KindModel {
+		return st.setAnnotationsForModel(ctx, db, id, toUpsert, toRemove)
+	}
+	return st.setAnnotationsForID(ctx, db, id, toUpsert, toRemove)
+}
+
+// setAnnotationsForID associates key/value pairs with the given ID. This is separate from the
+// setAnnotationsForModel because for non-model ID Kinds we need to find the uuid of the id before
+// we add an annotation in the corresponding annotation table.
+func (st *State) setAnnotationsForID(ctx context.Context, db database.TxnRunner, id annotations.ID,
+	toUpsert map[string]string, toRemove []string) error {
+
+	// Prepare query for getting the UUID of id.
+	kindQuery, kindQueryParam, err := uuidQueryForID(id)
+	if err != nil {
+		return errors.Annotatef(err, "preparing uuid retrieval query for ID: %q", id.Name)
+	}
+	kindQueryStmt, err := sqlair.Prepare(kindQuery, sqlair.M{})
+	if err != nil {
+		return errors.Annotatef(err, "preparing uuid retrieval query for ID: %q", id.Name)
+	}
+
+	// Prepare query for inserting and deleting annotations for id
+	var setAnnotationsQuery string
+	var deleteAnnotationsQuery string
+
+	var setAnnotationsStmt *sqlair.Statement
+	var deleteAnnotationsStmt *sqlair.Statement
+
+	// Set query
+	setAnnotationsQuery = setAnnotationQueryForID(id)
+	// Delete query
+	deleteAnnotationsQuery = deleteAnnotationsQueryForID(id, strings.Join(toRemove, ", "))
+
+	// Prepare sqlair statements
+	setAnnotationsStmt, err = sqlair.Prepare(setAnnotationsQuery, Annotation{}, sqlair.M{})
+	if err != nil {
+		return errors.Annotatef(err, "preparing set annotations query for ID: %q", id.Name)
+	}
+	deleteAnnotationsStmt, err = sqlair.Prepare(deleteAnnotationsQuery, Annotation{}, sqlair.M{})
+	if err != nil {
+		return errors.Annotatef(err, "preparing set annotations query for ID: %q", id.Name)
+	}
+
+	// Running transactions using sqlair statements
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		// We need to find the uuid of ID first, so looking it up
+		result := sqlair.M{}
+		err = tx.Query(ctx, kindQueryStmt, kindQueryParam).Get(result)
```

Comment:
It's interesting that you don't need `.Get(&result)` here.

Reply:
Wasn't intentional, regardless, we should pass a pointer 👍